Step3 로또 2등

goodbyeyo opened the pull request:
안녕하세요 🙂

상세한 피드백 항상 감사합니다 

이번에도 리뷰 잘 부탁드려요 🙏

goodbyeyo committed:
feat: Lotto 도메인 테스트 및 구현

goodbyeyo committed:
feat: 로또 번호, 티켓 발급 기능 구현 및 테스트

goodbyeyo committed:
refactor: Test 코드 내 도메인 로직 Class 분리 리팩토링

goodbyeyo committed:
feat: 당첨 로또 기능 구현

goodbyeyo committed:
feat: 로또 컨트롤러, 입출력 뷰 구현

goodbyeyo committed:
refactor: 피드백 반영, LottoTickets.calculateLottoRank 테스트 추가

goodbyeyo committed:
refactor: 피드백 반영
1. WinningLotto 제거하고 담당하던 책임 LottoTicket 으로 이동
2. LottoTicket 프로퍼티 변경 : List<Int> -> Set<LottoNumber>
3. LottoTicket 프로퍼티 기반으로 위임하도록 변경, 부생성자 추가
4. LottoTicket 정적팩토리메서드 추가

goodbyeyo committed:
doc: STEP3 로또2등 요구구사항 정리

goodbyeyo committed:
feat: 보너스번호 도메인 구현

goodbyeyo committed:
feat: 보너스 번호 일치여부 구현

goodbyeyo committed:
feat: 로또 2등 기능 구현
1. LottoRank : 로또 랭크 판단 조건, 상금 변경
2. LottoTicket : 당첨 로또 계산 위임(WinningLotto)
3. LottoTicket : data class 변경 (피드백 반영)
4. 당첨 로또 계산 시 보너스 번호가 포함된 WinningLotto 전달
5. 당첨 결과 출력 보너스 번호 일치 여부 출력
6. 테스트 코드 2등 로또 반영

vsh123 commented on src/main/kotlin/lotto/domain/BonusNumber.kt:
```diff
@@ -0,0 +1,17 @@
+package lotto.domain
+
+data class BonusNumber(private val number: LottoNumber) {
```
LottoNumber를 BonusNumber로 한번 더 wrapping하신 이유가 있을까요?? 두개의 역할이 달라보이지 않아서요!

vsh123 commented on src/main/kotlin/lotto/domain/WinningLotto.kt:
```diff
@@ -0,0 +1,21 @@
+package lotto.domain
+
+class WinningLotto(
+    private val winningLotto: LottoTicket,
+    number: String,
```
winningLotto는  LottoTicket으로 받고, bonusNumber는 String으로 받는 이유가 궁금합니다!! 
두개의 결을 맞춰주면 좋을 것 같아요~

vsh123 commented on src/main/kotlin/lotto/domain/BonusNumber.kt:
```diff
@@ -0,0 +1,17 @@
+            lottoTicket: LottoTicket,
+            number: LottoNumber,
+        ): BonusNumber {
+            require(!lottoTicket.contains(number)) { "보너스 번호는 로또 티켓의 번호와 중복될 수 없습니다" }
```
해당 validation을 체크하는게 BonusNumber의 역할일까요?

vsh123 commented on src/main/kotlin/lotto/domain/WinningLotto.kt:
```diff
@@ -0,0 +1,21 @@
+    private val bonusNumber: BonusNumber
+
+    init {
+        val intNumber = number.toIntOrNull() ?: throw IllegalArgumentException("보너스 번호가 유효하지 않습니다")
```
해당 validation에 대한 테스트가 추가되면 좋겠네요!

vsh123 reviewed (changes requested):
안녕하세요!
step3 너무잘 작성해주셨네요!! 💯 

몇가지 코멘트를 남겼으니 확인부탁드릴게요~!

goodbyeyo committed:
refactor: 피드백 반영
1. BonusNumber 제거하고 WinningLotto 가 직접 책임을 담당하도록 변경
2. WinningLotto 생성자 인자 팩토리 메서드(동반 객체 메서드) 구조로 전달
3. LottoNumber 정적 팩터리 메서드 추가

vsh123 commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -1,19 +1,17 @@
 package lotto.domain
 
 data class LottoNumber(val number: Int) {
```
private 생성자로 변환해도 괜찮겠네요 :)

vsh123 commented on src/main/kotlin/lotto/domain/LottoTicket.kt:
```diff
@@ -30,7 +30,7 @@ data class LottoTicket(private val numbers: Set<LottoNumber>) : Collection<Lotto
 
         fun from(numbers: Set<Int>): LottoTicket {
-            return LottoTicket(numbers.map { LottoNumber.from(it) }.toSet())
+            return LottoTicket(numbers.map { LottoNumber(it) }.toSet())
```
pool을 활용하면 더 좋을 것 같은데요! 어떠신가요?

vsh123 reviewed (approved):
안녕하세요! step3 리뷰 반영 잘해주셨네요!!

몇가지 코멘트를 남겼으니 다음 step에서 함께 고민해보시면 좋을 것 같습니다.

마지막까지 화이팅하세요!
